### PR TITLE
Features/unification artifacts

### PR DESCRIPTION
**Pull request by sfc-gh-astus**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
* Added support for glob pattern in artifact paths in snowflake.yml for Streamlit.
* Added support for glob pattern in artifact paths in snowflake.yml for Snowpark, requires ENABLE_SNOWPARK_GLOB_SUPPORT feature flag.
* Extracted artifacts common logic to common package.

### CODE REVIEW

**Comment by sfc-gh-astus** [comment]
Check if --project in streamlit works on this branch.

File: src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py [hunk 1]
```diff
@@ -59,17 +48,24 @@ class SnowparkEntityModel(EntityModelBase, ExternalAccessBaseModel, ImportsBaseM
...
-        for artefact in artifacts:
-            if isinstance(artefact, PathMapping):
-                _artifacts.append(artefact)
+        for artifact in artifacts:
+            if (
+                (isinstance(artifact, str) and glob.has_magic(artifact))
+                or (isinstance(artifact, PathMapping) and glob.has_magic(artifact.src))
+            ) and FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_disabled():
+                raise ValueError(
+                    "If you want to use glob patterns in artifacts, you need to enable the Snowpark new build feature flag (ENABLE_SNOWPARK_GLOB_SUPPORT=true)"
```
Comment (sfc-gh-pczajka) [inline src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py:63]:
```suggestion
                    "If you want to use glob patterns in artifacts, you need to enable the Snowpark new build feature flag (enable_snowpark_glob_support=true)"
```
I'm pretty sure that feature flags are read only as lowercase from `config.toml`. `SNOWFLAKE_CLI_FEATURES_ENABLE...` env variable also works

File: src/snowflake/cli/api/project/project_paths.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+class ProjectPaths:
+    project_root: Path
+
+    @property
+    def bundle_root(self) -> Path:
+        return bundle_root(self.project_root)
+
+    def remove_up_bundle_root(self) -> None:
+        if self.bundle_root.exists():
+            rmtree(self.bundle_root)
```
Comment (sfc-gh-pczajka) [inline src/snowflake/cli/api/project/project_paths.py:16]: Use `SecurePath.rmdir(recursive=True)` instead of `shutil.rmtree`

File: src/snowflake/cli/_plugins/streamlit/streamlit_entity.py [hunk 3]
```diff
@@ -70,29 +72,16 @@ def action_get_url(
...
+        return build_bundle(
+            self.root,
+            output_dir or bundle_root(self.root, "streamlit"),
+            [
+                PathMapping(
+                    src=artifact.src, dest=artifact.dest, processors=artifact.processors
+                )
+                for artifact in self._entity_model.artifacts
+            ],
+        )
```
Comment (sfc-gh-jsikorski) [inline src/snowflake/cli/_plugins/streamlit/streamlit_entity.py:84]: Bundle is an action, so it should not return bundle but rather copy the files

File: src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py (hunk 1 above)
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py:63]: Changed

File: src/snowflake/cli/api/project/project_paths.py (hunk 2 above)
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/project/project_paths.py:16]: Changed

File: src/snowflake/cli/api/artifacts/utils.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    absolute_src = resolve_without_follow(src)
+    if absolute_src.is_file():
+        delete(dst)
+        try:
+            os.symlink(absolute_src, dst)
+        except OSError:
+            ssrc.copy(dst)
+    else:
+        # 1. Create a new directory in the deploy root
+        dst.mkdir(exist_ok=True)
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/artifacts/utils.py:38]: Changed

File: src/snowflake/cli/api/artifacts/bundle_map.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    Computes the mapping between project directory artifacts (aka source artifacts) to their deploy root location
+    (aka destination artifact). This information is primarily used when bundling a native applications project.
+
+    :param project_root: The root directory of the project and base for all relative paths. Must be an absolute path.
+    :param deploy_root: The directory where artifacts should be copied to. Must be an absolute path.
+    """
+
+    def __init__(self, *, project_root: Path, deploy_root: Path):
+        # If a relative path ends up here, it's a bug in the app and can lead to other
+        # subtle bugs as paths would be resolved relative to the current working directory.
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/artifacts/bundle_map.py:45]: It is

File: src/snowflake/cli/api/artifacts/bundle_map.py [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    to specify "./" as its destination, or omit it (but only if the
+    file already lives in the project root).
+    """
+    return s.endswith("/")
+
+
+class BundleMap:
+    """
+    Computes the mapping between project directory artifacts (aka source artifacts) to their deploy root location
+    (aka destination artifact). This information is primarily used when bundling a native applications project.
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/artifacts/bundle_map.py:37]: Changed

File: src/snowflake/cli/api/artifacts/bundle_map.py [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+            )
+
+        return resolved_dest
+
+    def _canonical_src(self, src: Path) -> Path:
+        """
+        Returns the canonical version of a source path, relative to the project root.
+        """
+        absolute_src = self._absolute_src(src)
+        return absolute_src.relative_to(self._project_root)
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/artifacts/bundle_map.py:355]: This is nativeapp code moved to common package.

File: src/snowflake/cli/api/artifacts/bundle_map.py [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+class BundleMap:
+    """
+    Computes the mapping between project directory artifacts (aka source artifacts) to their deploy root location
+    (aka destination artifact). This information is primarily used when bundling a native applications project.
+
+    :param project_root: The root directory of the project and base for all relative paths. Must be an absolute path.
+    :param deploy_root: The directory where artifacts should be copied to. Must be an absolute path.
+    """
+
+    def __init__(self, *, project_root: Path, deploy_root: Path):
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/artifacts/bundle_map.py:43]: It will cause cascade :( , there will be a lot of changes in nativeapp code.

File: src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py [hunk 9]
```diff
@@ -57,17 +79,123 @@ def requirements(self) -> SecurePath:
...
-        return self.path_relative_to_root(Path("dependencies.zip"))
+    def bundle_root(self) -> Path:
+        return bundle_root(self.project_root, "snowpark")
 
 
 @dataclass(unsafe_hash=True)
 class Artefact:
     """Helper for getting paths related to given artefact."""
 
+    project_root: Path
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py:90]: Artefact logic is only for Snowpark, for me better fits here.

File: RELEASE-NOTES.md [hunk 10]
```diff
@@ -41,6 +41,8 @@
 * Add `snow spcs service metrics` command to fetch service metrics:
   * Supports filtering by service name, container name, instance ID, and time intervals (`--since`, `--until`).
   * Use `--all` to fetch all columns.
+* Added support for glob pattern in artifact paths in snowflake.yml for Streamlit.
+* Added support for glob pattern in artifact paths in snowflake.yml for Snowpark, requires ENABLE_SNOWPARK_GLOB_SUPPORT feature flag.
```
Comment (sfc-gh-turbaszek) [inline RELEASE-NOTES.md:45]: Is there any change to existing behavior?

Comment (sfc-gh-astus) [reply RELEASE-NOTES.md:45]: Streamlit is changed, Snowpark needs feature flag enabled, otherwise stays the same.

File: tests/snowpark/test_artifacts.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        assert {
+            "local_path": tmp_path / local_path,
+            "stage_path": "@MockDatabase.MockSchema.dev_deployment" + stage_path,
+        } in _extract_put_calls(mock_sm_put)
+
+
+@pytest.mark.parametrize(
+    "artifact, local_path, stage_path",
+    [
+        ("src", bundle_root / "src.zip", "/"),
```
Comment (sfc-gh-pczajka) [inline tests/snowpark/test_artifacts.py:116]: All paths in tests seem to start on `src/`. Do we support glob patterns like `**/snowpark/*.py`?

**Review by sfc-gh-jsikorski — APPROVED** [review]

File: tests/snowpark/test_artifacts.py (hunk 11 above)
Comment (sfc-gh-astus) [reply tests/snowpark/test_artifacts.py:116]: No, `**` introduce more complexity, so I decided to not add that. This PR is huge even without `**`.

Comment (sfc-gh-astus) [reply tests/snowpark/test_artifacts.py:116]: I added that to release notes.

**Review by sfc-gh-pczajka — DISMISSED** [review]

**Review by sfc-gh-jsikorski — DISMISSED** [review]

File: src/snowflake/cli/api/project/project_paths.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+from dataclasses import dataclass
+from pathlib import Path
+
+from snowflake.cli.api.secure_path import SecurePath
+
+
+@dataclass
+class ProjectPaths:
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/project/project_paths.py:8]: can we add docstring?

File: src/snowflake/cli/api/project/project_paths.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    @property
+    def bundle_root(self) -> Path:
+        return bundle_root(self.project_root)
+
+    def remove_up_bundle_root(self) -> None:
+        if self.bundle_root.exists():
+            SecurePath(self.bundle_root).rmdir(recursive=True)
+
+
+def bundle_root(root: Path, app_type: str | None = None) -> Path:
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/project/project_paths.py:20]: why is app_type in the path?

File: src/snowflake/cli/_plugins/snowpark/commands.py [hunk 14]
```diff
@@ -324,6 +328,9 @@ def build(
 
     anaconda_packages_manager = AnacondaPackagesManager()
 
+    # Clean up deploy root
+    project_paths.remove_up_bundle_root()
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/commands.py:332]: No, this clean up `deploy/bundle/snowpark` directory.

File: src/snowflake/cli/_plugins/snowpark/common.py [hunk 15]
```diff
@@ -214,6 +222,43 @@ def _standardize(packages: List[str]) -> Set[str]:
     return _standardize(old_dependencies) != _standardize(new_dependencies)
 
 
+def map_path_mapping_to_artifact(
+    project_paths: SnowparkProjectPaths, artifacts: List[PathMapping]
+) -> List[Artefact]:
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/common.py:227]: I will do this in next PR.

File: src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py [hunk 16]
```diff
@@ -59,17 +48,24 @@ class SnowparkEntityModel(EntityModelBase, ExternalAccessBaseModel, ImportsBaseM
...
     def _convert_artifacts(cls, artifacts: Union[dict, str]):
         _artifacts = []
-        for artefact in artifacts:
-            if isinstance(artefact, PathMapping):
-                _artifacts.append(artefact)
+        for artifact in artifacts:
+            if (
+                (isinstance(artifact, str) and glob.has_magic(artifact))
+                or (isinstance(artifact, PathMapping) and glob.has_magic(artifact.src))
+            ) and FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_disabled():
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_entity_model.py:61]: I check if artifact path has `glob magic`. Feature flag changes build and deploy commands behaviour, they use `deploy/bundle/...` directory.

File: src/snowflake/cli/api/project/project_paths.py (hunk 13 above)
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/project/project_paths.py:20]: To separate Streamlit and Snowpark artifacts, you can run 

```
snowpark build
streamlit deploy
snowpark deploy
```

File: src/snowflake/cli/api/project/schemas/entities/common.py [hunk 17]
```diff
@@ -162,3 +162,76 @@ def get_secrets_sql(self) -> str | None:
...
+                transformed_processors.append(ProcessorMapping(**input_processor))
+            else:
+                transformed_processors.append(input_processor)
+        return transformed_processors
+
+
+Artifacts = List[Union[PathMapping, str]]
+
+
+class ArtifactsBaseModel(EntityModelBase):
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/project/schemas/entities/common.py:215]: Changed to `EntityModelBaseWithArtifacts`

File: src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py [hunk 18]
```diff
@@ -57,17 +79,123 @@ def requirements(self) -> SecurePath:
...
+    ) -> None:
+        self.project_root = project_root
+        self.bundle_root = bundle_root
+        self.path = path
+        self.dest = dest
+        if self.dest and not self._is_dest_a_file() and not self.dest.endswith("/"):
+            self.dest = self.dest + "/"
+
+    @property
+    def _artefact_name(self) -> str:
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py:110]: Added

File: src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py [hunk 19]
```diff
@@ -57,17 +79,123 @@ def requirements(self) -> SecurePath:
...
+    def import_path(self, stage: FQN | str | None) -> str:
+        """Path for UDF/sproc imports clause."""
+        return self.upload_path(stage) + self._artefact_name
+
+    def _is_dest_a_file(self) -> bool:
+        if not self.dest:
+            return False
+        return re.search(r"\.[a-zA-Z0-9]{2,4}$", self.dest) is not None
+
+    def _path_until_asterisk(self) -> Path:
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py:175]: I don't understand the question.

File: src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py [hunk 20]
```diff
@@ -43,24 +44,20 @@ class StreamlitEntityModel(EntityModelBase, ExternalAccessBaseModel, ImportsBase
...
-        for artifact in self.artifacts:
-            if "*" in artifact.name:
-                continue
-            if not artifact.exists():
-                raise ValueError(
-                    f"Specified artifact {artifact} does not exist locally."
-                )
-
-        return self
+    @field_validator("artifacts")
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py:53]: What is ArtefactsBase? Can't find in code.

File: src/snowflake/cli/api/project/project_paths.py (hunk 12 above)
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/project/project_paths.py:8]: Added

File: src/snowflake/cli/api/project/project_paths.py (hunk 13 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/project/project_paths.py:25]: shouldn't we use entity ids instead, since you have multiple entities of each?

File: src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py:53]: sorry I meant ArtifactsBaseModel (now EntityModelBaseWithArtifacts)

File: src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py:180]: Sorry let me rephrase. This Artifact class, can't it also be used for Native Apps and other entities? If not, what's special about Snowpark artifacts?

[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py:180]: In Snowpark we build zip files for artifacts which are not single file. It is easier to import one zip file than multiple py files. Artifact class should be used only in Snowpark.

File: src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/_plugins/streamlit/streamlit_entity_model.py:53]: Good point! There is no problem with Snowpark, but Streamlit's artifacts were optional and when I changed to EntityModelBaseWithArtifacts now are required. It is BCR, I will talk with the team what we want to do with it.

File: src/snowflake/cli/api/project/project_paths.py (hunk 13 above)
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/project/project_paths.py:25]: There was that idea before but would introduce more complexity, so we decided to go simple way at the beginning.

**Review by sfc-gh-melnacouzi — DISMISSED** [review]

**Review by sfc-gh-pczajka — APPROVED** [review]

**Review by sfc-gh-melnacouzi — APPROVED** [review]

**Review by sfc-gh-jsikorski — APPROVED** [review]
